### DISCUSSION TITLE
Documentation: reverse proxying with NGINX needs proxy_cache_convert_head off

### DISCUSSION THREAD

**Discussion by spomata** [minio/minio#17097] [Q&A] [answered]
## Expected Behavior

Reverse proxying example configuration in the docs should include the option:
```
proxy_cache_convert_head off;
```

## Current Behavior

Without this option, I haven't been able to perform any file retrieval from MinIO buckets on my setup.
A "mc admin trace" before and after setting the option pointed me in the right direction: without the explicit declaration, HEAD requests would be converted to GET and fail with 403 - forbidden.

## Possible Solution

Improving documentation to help prospective users.

## Context

Setting up a MinIO SNSD development environment on RHEL 7 behind a NGINX reverse proxy.

## Regression

No

## Your Environment

* Version used (`minio --version`):
* RELEASE.2023-04-20T17-56-55Z
* Server setup and configuration:
RPM install
* Operating System and version (`uname -a`):
RHEL 7
`Linux host 5.4.225-1.el7.elrepo.x86_64 #1 SMP Fri Nov 25 11:40:22 EST 2022 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by harshavardhana** [comment]
> A "mc admin trace" before and after setting the option pointed me in the right direction: without the explicit declaration, HEAD requests would be converted to GET and fail with 403 - forbidden.

This is because you have NGINX caching enabled, this is not how by default NGINX behaves.

**Comment by spomata** [comment]
> > A "mc admin trace" before and after setting the option pointed me in the right direction: without the explicit declaration, HEAD requests would be converted to GET and fail with 403 - forbidden.
> 
> This is because you have NGINX caching enabled, this is not how by default NGINX behaves.

I doubt so, a grep of "proxy_cache" only yielded "proxy_cache off;" results.

Still, why not explicitly adding
```
proxy_cache off;
```
to the docs then?

**Comment by harshavardhana** [comment]
Please share your entire nginx.conf

**Comment by spomata** [comment]
```
server {
  listen *:443 ssl;
  server_name minio.MY.DOMAIN;

  ssl_certificate /etc/ssl/minio.MY.DOMAIN.crt;
  ssl_certificate_key /etc/ssl/minio.MY.DOMAIN.key;

   # Allow special characters in headers
   ignore_invalid_headers off;
   # Allow any size file to be uploaded.
   # Set to a value such as 1000m; to restrict file size to a specific value
   client_max_body_size 0;
   # Disable buffering
   proxy_buffering off;
   proxy_request_buffering off;

   location / {
      proxy_set_header Host $http_host;
      proxy_set_header X-Real-IP $remote_addr;
      proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header X-Forwarded-Proto $scheme;

      proxy_connect_timeout 300;
      # Default is HTTP/1, keepalive is only enabled in HTTP/1.1
      proxy_http_version 1.1;
      proxy_set_header Connection "";
      chunked_transfer_encoding off;

      proxy_pass http://localhost:9000/; # This uses the upstream directive definition to load balance
   }
}
```

That's the server snippet that I basically copy-pasted from MinIO documentation before adding the `proxy_cache_convert_head off;` snippet. I also double checked the full config and effectively the proxy_cache is enabled in the http section since we are using Omnibus GitLab nginx as a reverse proxy.

I also found issue #4120 which mentions a similar issue, hence I'd really suggest adding a note regarding proxy_cache in the documentation, or disable it altogether or just adding a `proxy_cache_convert_head off;` by default.

**Reply by AlexZIX** [reply]
I use following setting for nginx when use it without caching as front to minio. It works fine without proxy_cache_convert_head.

proxy_set_header			Host				$http_host;
proxy_set_header			X-Forwarded-By		$server_addr:$server_port;
proxy_set_header			X-Forwarded-Proto	$scheme;

proxy_http_version			1.1;
proxy_set_header			Upgrade				$http_upgrade;
proxy_set_header			Connection			$connection_upgrade;
proxy_cache_bypass			$http_upgrade;

proxy_buffering				off;
proxy_ignore_client_abort	off;
proxy_request_buffering 	off;

add_header					X-ReverseProxy		"BAZIS-RP1";

proxy_connect_timeout		10s;
proxy_send_timeout          600s;
proxy_read_timeout          600s;
proxy_ssl_verify			off;
proxy_ssl_server_name 		on;
resolver					192.168.158.99 192.168.157.229 ipv6=off;

set_real_ip_from 			127.0.0.1/32;
real_ip_header 				proxy_protocol;
proxy_set_header 			X-Real-IP 			$proxy_protocol_addr;
proxy_set_header 			X-Fowarded-For 		$proxy_protocol_addr;

**Reply by spomata** [reply]
I don't disagree, it does work without the proxy_cache_convert_head . However if your nginx installation is more complex and includes a system-wide cache enabling, it will go unnoticed on the individual server section and MinIO won't work, requiring time to debug server-side and notice a HEAD request being treated as GET.

Hence, since it is **vital** for MinIO to turn caching off, I don't get why not explicitly adding it to the server section.

**Reply by AlexZIX** [reply]
I just use the dedicated config with caching enabled and dedicated domain:
s3.bazissoft.ru - proxy without caching
cdn.bazissoft.ru - proxy with caching

But both proxying to same Minio instance. I'm not sure that it's a good idea to cache all S3 requests include listobjects for example. So in my case we cache only static content.

**Reply by spomata** [reply]
So how do you selectively enable caching for **get** requests only in nginx?

**Reply by AlexZIX** [reply]
Nginx have special directive for this:

proxy_cache_methods GET;

**Reply by AlexZIX** [reply, marked as answer]
This is full config with caching enabled. May be it will be helpful for you.

proxy_set_header				Host				$http_host;
proxy_set_header				X-Forwarded-By		$server_addr:$server_port;
proxy_set_header				X-Forwarded-For		$remote_addr;
proxy_set_header				X-Forwarded-Proto	$scheme;
proxy_set_header 				X-Real-IP 			$remote_addr;

proxy_cache 					nginx_cache;
proxy_cache_key 				$request_method$scheme$host$request_uri;

proxy_cache_convert_head		off; 

proxy_cache_revalidate			on;
proxy_cache_min_uses 			3;
proxy_cache_use_stale 			error timeout updating http_500 http_502 http_503 http_504;
proxy_cache_background_update 	on;
proxy_cache_lock				on;

proxy_ignore_headers 			Cache-Control;
proxy_cache_valid 				5m;
proxy_cache_methods 			GET;

proxy_request_buffering 		off;

add_header						X-ReverseProxy		$proxy_server;
add_header						X-ProxyCache		"Enabled";
add_header 						X-Cache-Status 		$upstream_cache_status;

**Reply by Kinuseka** [reply]
Disabling proxy was way too slow in my setup. Thank you for sharing this! my endpoint loads blazingly fast now.
